### PR TITLE
Have a list backed by network only, using Flow

### PR DESCRIPTION
**Pull request by florina-muntenescu**

### CODE REVIEW

File: app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt [hunk 1]
```diff
@@ -17,60 +17,75 @@
...
-class GithubRepository(
-    private val service: GithubService,
-    private val cache: GithubLocalCache
-) {
+@FlowPreview
+@ExperimentalCoroutinesApi
+class GithubRepository(private val service: GithubService) {
+
+    // keep the list of responses
+    private val inMemoryCache = ConflatedBroadcastChannel<RepoSearchResult>()
```
Comment (ChrisCraik) [inline app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt]: This seems a little strange to me, having a channel like this as a cache, is this an idiomatic way to use it? (I'm just not familiar enough to say)

File: app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt [hunk 2]
```diff
@@ -17,60 +17,75 @@
...
  */
-class GithubRepository(
-    private val service: GithubService,
-    private val cache: GithubLocalCache
-) {
+@FlowPreview
+@ExperimentalCoroutinesApi
+class GithubRepository(private val service: GithubService) {
+
+    // keep the list of responses
```
Comment (manuelvicnt) [inline app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt]: In reality, this doesn't keep the list of responses anymore. Whenever you start listening to this channel you'll get the last result emitted by the producer and subsequent emissions.

File: app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt [hunk 3]
```diff
@@ -17,60 +17,75 @@
...
-    private val networkErrors = MutableLiveData<String>()
-
     // avoid triggering multiple requests in the same time
     private var isRequestInProgress = false
 
     /**
      * Search repositories whose names match the query.
      */
-    fun search(query: String): RepoSearchResult {
+    suspend fun search(query: String): Flow<RepoSearchResult> {
```
Comment (manuelvicnt) [inline app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt]: This method is a bit confusing for me as the name is `search` (i.e. perform a search operation) but also returns a `inMemoryCache` as Flow with which you will receive the searched item + other updates that can happen.

Is that intended? Something is wrong with that for me, I think it's the name of the function, that should say more

File: app/src/main/java/com/example/android/codelabs/paging/ui/SearchRepositoriesViewModel.kt [hunk 4]
```diff
@@ -16,32 +16,29 @@
...
 
     private val queryLiveData = MutableLiveData<String>()
-    private val repoResult: LiveData<RepoSearchResult> = Transformations.map(queryLiveData) {
-        repository.search(it)
-    }
-
-    val repos: LiveData<List<Repo>> = Transformations.switchMap(repoResult) { it -> it.data }
-    val networkErrors: LiveData<String> = Transformations.switchMap(repoResult) { it ->
-        it.networkErrors
+    val repoResult: LiveData<RepoSearchResult> = Transformations.switchMap(queryLiveData) {
```
Comment (manuelvicnt) [inline app/src/main/java/com/example/android/codelabs/paging/ui/SearchRepositoriesViewModel.kt]: I think there are some available extension functions to do queryLiveData.switchMap { 
   liveData { ... }
}

File: app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt (hunk 1 above)
Comment (florina-muntenescu) [reply app/src/main/java/com/example/android/codelabs/paging/data/GithubRepository.kt]: @manuelvicnt said this is ok.

File: app/src/main/java/com/example/android/codelabs/paging/api/GithubService.kt [hunk 5]
```diff
@@ -38,44 +36,30 @@ private const val IN_QUALIFIER = "in:name,description"
...
-            }
-    )
+    val response = service.searchRepos(apiQuery, page, itemsPerPage)
+    if (response.isSuccessful) {
+        Log.d(TAG, "got a response $response")
+        if (response.isSuccessful) {
+            val repos = response.body()?.items ?: emptyList()
+            return RepoSearchResult(repos)
+        } else {
+            return RepoSearchResult(emptyList(), response.message() ?: "Unknown error")
```
[the remark this answers is not in the document]
Comment (florina-muntenescu) [reply app/src/main/java/com/example/android/codelabs/paging/api/GithubService.kt]: Using exceptions but not a sealed class as it's not a success or error case - i want to be able to show the list even when i have an error

File: app/src/main/java/com/example/android/codelabs/paging/ui/SearchRepositoriesActivity.kt [hunk 6]
```diff
@@ -62,13 +66,13 @@ class SearchRepositoriesActivity : AppCompatActivity() {
...
     private fun initAdapter() {
         list.adapter = adapter
-        viewModel.repos.observe(this, Observer<List<Repo>> {
-            Log.d("Activity", "list: ${it?.size}")
-            showEmptyList(it?.size == 0)
-            adapter.submitList(it)
-        })
-        viewModel.networkErrors.observe(this, Observer<String> {
-            Toast.makeText(this, "\uD83D\uDE28 Wooops $it", Toast.LENGTH_LONG).show()
+        viewModel.repoResult.observe(this, Observer<RepoSearchResult> {
```
[the remark this answers is not in the document]
Comment (florina-muntenescu) [reply app/src/main/java/com/example/android/codelabs/paging/ui/SearchRepositoriesActivity.kt]: there is!

**Pull request closed without merging**